x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java

```diff
@@ -0,0 +1,143 @@
+    }
+
+    @Override
+    protected RestoreSnapshotResponse read(StreamInput in) throws IOException {
+        return new RestoreSnapshotResponse(in);
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(MountSearchableSnapshotRequest request, ClusterState state) {
+        // The restore action checks the cluster blocks.
+        return null;
+    }
+
+    /**
+     * Return the index settings required to make a snapshot searchable
+     */
+    private static Settings getIndexSettings(String repoName, SnapshotId snapshotId, IndexId indexId) {
+        return Settings.builder()
+            .put(SearchableSnapshots.SNAPSHOT_REPOSITORY_SETTING.getKey(), repoName)
+            .put(SearchableSnapshots.SNAPSHOT_SNAPSHOT_NAME_SETTING.getKey(), snapshotId.getName())
+            .put(SearchableSnapshots.SNAPSHOT_SNAPSHOT_ID_SETTING.getKey(), snapshotId.getUUID())
+            .put(SearchableSnapshots.SNAPSHOT_INDEX_ID_SETTING.getKey(), indexId.getId())
+            .put(INDEX_STORE_TYPE_SETTING.getKey(), SearchableSnapshots.SNAPSHOT_DIRECTORY_FACTORY_KEY)
+            .put(IndexMetaData.SETTING_BLOCKS_WRITE, true)
+            .build();
+    }
+
+    @Override
+    protected void masterOperation(Task task, final MountSearchableSnapshotRequest request, final ClusterState state,
+                                   final ActionListener<RestoreSnapshotResponse> listener) {
+        final String repoName = request.repositoryName();
+        final String snapName = request.snapshotName();
+        final String indexName = request.snapshotIndexName();
+
+        // Retrieve IndexId and SnapshotId instances, which are then used to create a new restore
+        // request, which is then sent on to the actual snapshot restore mechanism
+        final Repository repository = repositoriesService.repository(repoName);
+        final StepListener<RepositoryData> repositoryDataListener = new StepListener<>();
+        repository.getRepositoryData(repositoryDataListener);
+        repositoryDataListener.whenComplete(repoData -> {
+            final Map<String, IndexId> indexIds = repoData.getIndices();
+            if (indexIds.containsKey(indexName) == false && indexIds.get(indexName) != null) {
+                throw new IndexNotFoundException("index [" + indexName + "] not found in repository [" + repoName + "]");
+            }
+            final IndexId indexId = indexIds.get(indexName);
+
+            final Optional<SnapshotId> matchingSnapshotId = repoData.getSnapshotIds().stream()
+                .filter(s -> snapName.equals(s.getName())).findFirst();
+            if (matchingSnapshotId.isEmpty()) {
+                throw new ElasticsearchException("snapshot [" + snapName + "] not found in repository [" + repoName + "]");
+            }
+            final SnapshotId snapshotId = matchingSnapshotId.get();
+
+            // NORELEASE TODO should we ensure that the IDs we just obtained match the ones that we ultimately restore?
```

Comment:
This is a good observation. I don't have a strong opinion about it as the `Directory` implementation ensures (from index settings) that we are restoring the snapshot id/index id discovered here; but I agree that a different snapshot and/or index could be resolved later by the transport restore action itself (and if that index has a different number of shards, boom). I see this as a low risk that could maybe be worked around by adding snapshot it and index id to the restore request and check them (if present) during restore?

Reply:
Yes, it'd involve adding extra data to the restore so that we can re-validate them later. I think we need to do this TBH, or else we have no guarantees that the index metadata we restore corresponds with the snapshot.

I expanded the comment in 414e830129249bff43ccec131543f939c3a1b8b7 but we can address this later.

Reply:
:+1: